### PR TITLE
[smart_holder] Change `throw_if_uninitialized_or_disowned_holder()` to also show C++ type name.

### PR DESCRIPTION
**Pull request by rwgk**
## Description
To aid debugging.

## Suggested changelog entry:

```rst

```

### CODE REVIEW

**Comment by rwgk** [comment]
@karagog could you please help reviewing this PR?

File: include/pybind11/detail/smart_holder_type_casters.h [hunk 1]
```diff
@@ -694,17 +694,22 @@ struct smart_holder_type_caster_load {
     holder_type &holder() const { return load_impl.loaded_v_h.holder<holder_type>(); }
 
     // have_holder() must be true or this function will fail.
-    void throw_if_uninitialized_or_disowned_holder() const {
+    void throw_if_uninitialized_or_disowned_holder(const char *typeid_name) const {
+        static const std::string missing_value_msg = "Missing value for wrapped C++ type `";
         if (!holder().is_populated) {
-            pybind11_fail("Missing value for wrapped C++ type:"
-                          " Python instance is uninitialized.");
+            throw value_error(missing_value_msg + clean_type_id(typeid_name)
```
Comment (karagog) [inline include/pybind11/detail/smart_holder_type_casters.h:700]: I dont' suppose you can use f-strings?

File: tests/test_class_sh_property.py [hunk 2]
```diff
@@ -9,23 +9,19 @@
 
 @pytest.mark.xfail("env.PYPY", reason="gc after `del field` is apparently deferred")
 @pytest.mark.parametrize("m_attr", ["m_valu_readonly", "m_valu_readwrite"])
-def test_valu_getter(msg, m_attr):
+def test_valu_getter(m_attr):
```
Comment (karagog) [inline tests/test_class_sh_property.py:12]: nit: this function name is apparently misspelled

**Review by karagog — APPROVED** [review]
LGTM

File: include/pybind11/detail/smart_holder_type_casters.h (hunk 1 above)
Comment (rwgk) [reply include/pybind11/detail/smart_holder_type_casters.h:700]: Interesting idea, but no. We don't have f-strings in C++ (I never thought about that before, not sure how feasible that is), and compared to relying on `std::string` `operator+`, the overhead of using the Python C API is not something I'd want to have in an error handling code path like here.

File: tests/test_class_sh_property.py (hunk 2 above)
Comment (rwgk) [reply tests/test_class_sh_property.py:12]: It's intentional, see comment at top of this file, which points here:

https://github.com/pybind/pybind11/blob/feed8b15b581b317c92f00777c15d8ed5fe8a56f/tests/test_class_sh_property.cpp#L1-L4

(The original idea was actually to make [this code in test_class_sh_basic.cpp](https://github.com/pybind/pybind11/blob/feed8b15b581b317c92f00777c15d8ed5fe8a56f/tests/test_class_sh_basic.cpp#L70-L81) more readable.)

**Comment by rwgk** [comment]
Thanks for the review!

**Pull request merged**